Improve doc (#45)
cowprotocol/cow-sdk: README.md

Posting orders is a three steps process:

- 1. **Get Market Pricea**: Fee & Price
- 2. **Sign the order**: Using off-chain signing or Meta-transactions
- 3. **Post the signed order to the API**: So, the order becomes `OPEN`

### Enable tokens (token approval)
Before we jump to the fun part of creating an order, let's briefly mention some small disclamer. 

The protocol requires you to approve (enable) the sell token before the order can be considered.

This approval is something that is done once, and after all order creation can be done for free using offchain signing.

> For more details see https://docs.cow.fi/tutorials/how-to-submit-orders-via-the-api/1.-set-allowance-for-the-sell-token

### Instantiate SDK with a signer
Before you can sign any transaction, you have to instantiate the SDK with a [Ethers.JS signer](https://docs.ethers.io/v5/api/signer/):

```js
import { Wallet } from 'ethers'
import { CowSdk, OrderKind } from '@cowprotocol/cow-sdk'

const mnemonic = 'fall dirt bread cactus...'
const wallet = Wallet.fromMnemonic(mnemonic)
const cowSdk = new CowSdk(
  4, {            // Leaving chainId empty will default to MAINNET

>   signer: wallet  // Provide a signer, so you can sign order

Reviewer: ```suggestion
    signer: wallet  // Provide a signer, so you can sign order
```

})

const { sellToken, buyToken, validTo, buyAmount, sellAmount, receiver, feeAmount } = quoteResponse.quote
```

### STEP 2: Sign the order
Once you know the price and fee, we can create the order and sign it:

  * Technically the order is just a signed message with your intent to trade, and contains your `Limit Price` and `Fee`.
  * As explained before, you can choose your `Limit Price`, but some general approach is to take the current Market Price
  and apply some slippage tolerance to it. `Received Amount = Expected Amount * (1 - Slippage Tolerance)`
  * The SDK will provide an easy way to sign orders given the raw data

```js
const { sellToken, buyToken, validTo, buyAmount, sellAmount, receiver, feeAmount } = quoteResponse.quote

// Prepare the RAW order
const order = {
  kind: OrderKind.SELL, // SELL / BUY  
  receiver, // Your account or any other
  sellToken,
  buyToken,

>   partiallyFillable: false, // (true would be for a "Fill or Kill" order, "false" for allowing "Partial execution" which is not supported yet)

Reviewer: It's the other way around

```suggestion
  partiallyFillable: false, // ("false" is for a "Fill or Kill" order, "true" for allowing "Partial execution" which is not supported yet)
```

  // Deadline
  validTo,

  // Limit Price
  //    You can apply some slippage tolerance here to make sure the trade is executed. 
  //    CoW protocol protects from MEV, so it can work with higher slippages
  sellAmount,
  buyAmount, 

  // Use the fee you received from the API
  feeAmount,

  // The appData allows you to attach arbitrary information (meta-data) to the order. Its explained in their own section. For now, you can use this 0x0 value
  appData: '0x0000000000000000000000000000000000000000000000000000000000000000'
}

// Sign the order
const signedOrder = await cowSdk.signOrder(order)
```

At this point, you have a signed order. So next step will be to post it to the API so it's considered by the solvers and executed.

  
## STEP 3: **Post the signed order to the API**: 
Once you have a signed order, last step is to send it to the API.
  * The API will accept the order if its correctly signed, the deadline is correct, and the fee is enough to settle it

>   *  Once accepted, the order will be `OPEN` until the specified `validTo` date (expiration)

Reviewer: ```suggestion
  * Once accepted, the order will be `OPEN` until the specified `validTo` date (expiration)
```

>   * The possible outcomes once is accepted is:

Reviewer: ```suggestion
  * The possible outcomes once accepted is:
```

>       * The order is `EXECUTED`: you will pay the signed fee, and get at least the `buyAmount` tokens you specified, although you will probably get more! (you will probably get a so-called **Surplus**).

Reviewer: Should we make clear this is in regards to a sell order?
Should we describe how it would work for buy orders?

)
```

> ## Create a meta-data document for attaching into an order

Reviewer: ```suggestion
## Create a meta-data document for attaching to an order
```

The most important thing to define in the meta-data is the name of your app, so the order-flow can be credited to it.

```js
const appDataDoc = cowSdk.metadataApi.generateAppDataDoc({}, {
  appCode: 'YourApp'
})
```

This will create a document similar to:
```json
{
  "version": "0.1.0",
  "appCode": "YourApp",
  "metadata": {},
} 
```

After creating the most basic document, you can see how to attach additional meta-data items.

For example, you could give information about who reffered the user creating the order.

```js
const appDataDoc = cowSdk.metadataApi.generateAppDataDoc(
  {
    referrer: {
      address: '0x1f5B740436Fc5935622e92aa3b46818906F416E9',
      version: '0.1.0',
    },
  },
  {
    appCode: 'YourApp',
  }
)
```

This will create a document similar to:

```json
{
    "version": "0.1.0",
    "appCode": "YourApp",
    "metadata": {
      "referrer": {
        "address": "0x1f5B740436Fc5935622e92aa3b46818906F416E9",
        "version": "0.1.0",
      },
    },
}
```

For a complete list of meta-data that can be attach check [@cowprotocol/app-data](https://github.com/cowprotocol/app-data)

## Get the AppData Hex

> The `AppData` Hex points to a IPFS document with the meta-data attached to the order.

Reviewer: ```suggestion
The `AppData` Hex points to an IPFS document with the meta-data attached to the order.
```

You can calculate the `AppData` Hex, and its corresponding `cidV0` using the SDK:

```js
const { appDataHash, cidv0 } = await cowSdk.metadataApi.calculateAppDataHash(appDataDoc)
```

Note how this operation is deterministic, so given the same document, it will always generate the same hash. 

> This method can be used to calculate the actual hash before posting the order, or even uploading the document to IPFS.

Reviewer: Actually you must have the hash before posting the order

```suggestion
This method can be used to calculate the actual hash before uploading the document to IPFS.
```

## Get meta-data document uploaded to IPFS from AppData
Given the `AppData` of a document that has been uploaded to IPFS, you can easily retrieve the document.

```js
const appDataDoc = await cowSdk.metadataApi.decodeAppData('0x5ddb2c8207c10b96fac92cb934ef9ba004bc007a073c9e5b13edc422f209ed80')
```

This will return a document similar to:

```json
{
    "version": "0.1.0",
    "appCode": "YourApp",
    "metadata": {
      "referrer": {
        "address": "0x1f5B740436Fc5935622e92aa3b46818906F416E9",
        "version": "0.1.0",
      },
    },
}
```

## Upload document to IPFS
The SDK uses Pinata to upload it to IPFS, so you will need an API Key in order to upload it using the SDK.

Alternativelly, you can upload the document on your own using any other service.

```js

> // Make sure yuou provide the IPFS params when instanciating the SDK

Reviewer: ```suggestion
// Make sure you provide the IPFS params when instantiating the SDK
```
